src/main/java/com/linkedin/kmf/services/ClusterTopicManipulationService.java

```diff
@@ -0,0 +1,242 @@
+      // unaccessed.
+    }
+
+    /**
+     * When an object implementing interface <code>Runnable</code> is used
+     * to create a thread, starting the thread causes the object's
+     * <code>run</code> method to be called in that separately executing
+     * thread.
+     * <p>
+     * The general contract of the method <code>run</code> is that it may
+     * take any action whatsoever.
+     *
+     * @see     Thread#run()
+     */
+    @Override
+    public void run() {
+      try {
+        ClusterTopicManipulationService.this.createDeleteClusterTopic();
+      } catch (Exception e) {
+        LOGGER.error("{} {} failed to run createDeleteClusterTopic()", _configDefinedServiceName,
+            ClusterTopicManipulationService.this.getClass().getSimpleName(), e);
+      }
+    }
+  }
+
+  private void createDeleteClusterTopic() {
+
+    if (_isOngoingTopicCreationDone) {
+
+      int random = ThreadLocalRandom.current().nextInt();
+      _currentlyOngoingTopic = XinfraMonitorConstants.TOPIC_MANIPULATION_SERVICE_TOPIC + Math.abs(random);
+
+      try {
+        int brokerCount = _adminClient.describeCluster().nodes().get().size();
+        _createTopicsResult = _adminClient.createTopics(Collections.singleton(
+            new NewTopic(_currentlyOngoingTopic, XinfraMonitorConstants.TOPIC_MANIPULATION_TOPIC_NUM_PARTITIONS,
+                (short) brokerCount)));
+
+        _totalPartitions.set(brokerCount * XinfraMonitorConstants.TOPIC_MANIPULATION_TOPIC_NUM_PARTITIONS);
+
+        _isOngoingTopicCreationDone = false;
+        LOGGER.trace("Initiated a new topic creation. topic information - topic: {}, cluster broker count: {}",
+            _currentlyOngoingTopic, brokerCount);
+      } catch (InterruptedException | ExecutionException e) {
+        LOGGER.error("Exception occurred while retrieving the brokers count: ", e);
+      }
+    }
+
+    try {
+      LOGGER.trace("cluster id: {}", _adminClient.describeCluster().clusterId().get());
+      Collection<Node> brokers = _adminClient.describeCluster().nodes().get();
+      if (_createTopicsResult.all().isDone() && doesClusterContainTopic(_currentlyOngoingTopic, brokers,
+          _adminClient)) {
+        _adminClient.deleteTopics(Collections.singleton(_currentlyOngoingTopic)).all();
+        LOGGER.trace("clusterTopicManipulationServiceRunnable: Initiated topic deletion on {}.",
+            _currentlyOngoingTopic);
+        _isOngoingTopicCreationDone = true;
+        LOGGER.trace("{}-clusterTopicManipulationServiceRunnable successful!", this.getClass().getSimpleName());
+      }
+    } catch (InterruptedException | ExecutionException e) {
+      LOGGER.error("Exception occurred while creating cluster topic in {}: ", _configDefinedServiceName, e);
+    }
+  }
+
+  private boolean doesClusterContainTopic(String topic, Collection<Node> brokers, AdminClient adminClient)
+      throws ExecutionException, InterruptedException {
+    for (Node broker : brokers) {
```

Comment:
This for-loop is not necessary. The `LOGGER.trace()` can be merged to the below for-loop.
I guess that `adminClient.describeLogDirs` is expensive RPC. In this for-loop, the results from this RPC are ignored most of the time and the below for-loop does it again but actually uses the results. So, please merge these 2 for-loops and make sure `adminClient.describeLogDirs` is called once for every broker only.

Reply:
👍

Reply:
`adminClient.describeLogDirs(Collections.singleton(broker.id())).all().get()` is still invoked twice back to back right? I think even though `LOGGER.tracd(...)` usually does not log anything. But `adminClient.describeLogDirs(Collections.singleton(broker.id())).all().get()` is still going to be invoked there and it's just it's results are not logged.

You can call `adminClient.describeLogDirs(Collections.singleton(broker.id())).all().get()` only once and use its results for logging and later process